web/src/compose_ui.js

```diff
@@ -375,6 +385,122 @@ export function format_text($textarea, type, inserted_content) {
 
+    const section_off_selected_lines = () => {
+        // Divide all lines of text (separated by `\n`) into those entirely or
+        // partially selected, and those before and after these selected lines.
+        const before = text.slice(0, range.start);
+        const after = text.slice(range.end);
+        let separating_new_line_before = false;
+        let closest_new_line_beginning_before_index;
+        if (before.includes("\n")) {
+            separating_new_line_before = true;
+            closest_new_line_beginning_before_index = before.lastIndexOf("\n");
+        } else {
+            separating_new_line_before = false;
+            // The beginning of the entire text acts as a new line.
+            closest_new_line_beginning_before_index = -1;
+        }
+        let separating_new_line_after = false;
+        let closest_new_line_char_after_index;
+        if (after.includes("\n")) {
+            separating_new_line_after = true;
+            closest_new_line_char_after_index =
+                after.indexOf("\n") + before.length + selected_text.length;
+        } else {
+            separating_new_line_after = false;
+            // The end of the entire text acts as a new line.
+            closest_new_line_char_after_index = text.length;
+        }
+        // selected_lines neither includes the `\n` character that marks it's
+        // beginning (which exists if there are before_lines) nor the one
+        // that marks it's end (which exists if there are after_lines).
+        const selected_lines = text.slice(
+            closest_new_line_beginning_before_index + 1,
+            closest_new_line_char_after_index,
+        );
+        // before_lines excludes the `\n` character that separates it from selected_lines.
+        const before_lines = text.slice(0, Math.max(0, closest_new_line_beginning_before_index));
+        // after_lines excludes the `\n` character that separates it from selected_lines.
+        const after_lines = text.slice(closest_new_line_char_after_index + 1);
+        return {
+            before_lines,
+            separating_new_line_before,
+            selected_lines,
+            separating_new_line_after,
+            after_lines,
+        };
+    };
+
+    const format_list = (type) => {
+        let is_marked;
+        let mark;
+        let strip_marking;
+        if (type === "bulleted") {
+            is_marked = bulleted_numbered_list_util.is_bulleted;
+            mark = (line) => "- " + line;
+            strip_marking = bulleted_numbered_list_util.strip_bullet;
+        } else {
+            is_marked = bulleted_numbered_list_util.is_numbered;
+            mark = (line, i) => i + 1 + ". " + line;
+            strip_marking = bulleted_numbered_list_util.strip_numbering;
+        }
+        // Following behaviour (specifically, the toggling of complete lines even when they are
+        // partially selected (and just selecting the newline character after a line counts as
```

Comment:
Your comment here has a parenthetical inside a parenthetical, but rather than fixing the punctuation, can you rewrite this so its a little easier to follow?

Reply:
On second thought, we don't want to risk this comment becoming nonsensical if GitHub changes their behavior. So consider striking this, or rewriting just to explain what the code is doing.

Reply:
Retained just the 1st point about considering complete lines even on partial selection, since all other points already have comments explaining them.